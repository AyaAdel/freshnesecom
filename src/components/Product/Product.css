.product-item {
  background: #ffffff;
  border: 1px solid #d1d1d1;
  border-radius: 12px;
  padding: 16px;
}

.product-item__image-wrapper {
  position: relative;
  margin-bottom: 16px;
  height: 180px;
}

.product-item__image {
  height: 180px;
  width: 100%;
  object-fit: cover;
  border-radius: 12px;
}

.product-item__discount-percent {
  position: absolute;
  top: 12px;
  left: 12px;
  background: #f4f8ec;
  border-radius: 12px;
  font-weight: 600;
  font-size: 12px;
  line-height: 18px;
  color: #6a983c;
  padding: 0 8px;
}

.product-item__name {
  font-weight: 500;
  font-size: 15px;
  line-height: 22px;
  color: #151515;
  margin: 0;
  text-transform: capitalize;
  margin-bottom: 4px;
}

.product-item__description {
  font-size: 12px;
  line-height: 16px;
  color: #575757;
  margin: 0 0 16px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.product-item__Action {
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.product-item__price {
  display: flex;
  flex-direction: column;
}

.product-item__current-price {
  text-transform: uppercase;
  font-weight: 600;
  font-size: 18px;
  line-height: 27px;
  color: #151515;
}

.product-item__previous-price {
  font-weight: 600;
  font-size: 12px;
  line-height: 18px;
  color: #a9a9a9;
  text-decoration: line-through;
}

.btn {
  background: #6a983c;
  border: 2px solid #46760a;
  border-radius: 12px;
  font-weight: 700;
  font-size: 15px;
  line-height: 22px;
  color: #ffffff;
  padding: 6.5px 12px;
}

.btn:hover {
  cursor: pointer;
}
