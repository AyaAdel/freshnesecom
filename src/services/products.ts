import { http } from 'utils/http';

export const getAllProducts = (limit = 9, skip = 0) =>
  http(`https://dummyjson.com/products?limit=${limit}&skip=${skip}`);

export const allProductsCategories = () =>
  http('https://dummyjson.com/products/categories');

export const searchProducts = (productName: string) =>
  http(`https://dummyjson.com/products/search?q=${productName}`);

export const getProductsOfCategory = (category: string) =>
  http(`https://dummyjson.com/products/category/${category}`);
