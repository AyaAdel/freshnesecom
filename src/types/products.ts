export type ProductItem = {
  brand: string;
  category: string;
  description: string;
  discountPercentage: number;
  id: number;
  images: string[];
  price: number;
  rating: number;
  stock: number;
  thumbnail: string;
  title: string;
};

export type AllProducts = {
  limit: number;
  products: ProductItem[];
  skip: number;
  total: number;
};
