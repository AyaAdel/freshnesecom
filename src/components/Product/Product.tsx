import type { ProductItem } from 'types/products';
import { useAppDispatch } from 'app/hooks';
import type { AppDispatch } from 'app/store';
import { addToCart } from 'features/cart/cartSlice';

import './Product.css';

export default function Product({
  id,
  title,
  thumbnail,
  price,
  discountPercentage,
  description,
}: ProductItem) {
  const dispatch: AppDispatch = useAppDispatch();

  const addProductToCart = () => dispatch(addToCart());

  return (
    <div className="product-item" data-testid={`product${id}`}>
      <div className="product-item__image-wrapper">
        <img className="product-item__image" src={thumbnail} />
        <span className="product-item__discount-percent">{`-${discountPercentage}%`}</span>
      </div>
      <h5 className="product-item__name">{title}</h5>
      <p className="product-item__description">{description}</p>
      <div className="product-item__Action">
        <div className="product-item__price">
          <span className="product-item__current-price">{`${price} usd`}</span>
        </div>
        <button className="btn" onClick={addProductToCart}>
          Buy now
        </button>
      </div>
    </div>
  );
}
