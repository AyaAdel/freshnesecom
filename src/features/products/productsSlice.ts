import { createSlice } from '@reduxjs/toolkit';
import { AllProducts } from 'types/products';

const initialState: AllProducts = {} as AllProducts;

export const productsSlice = createSlice({
  name: 'products',
  initialState,
  reducers: {
    loadProducts: (state, action) => (state = action.payload),
  },
});

export const { loadProducts } = productsSlice.actions;

export default productsSlice.reducer;
