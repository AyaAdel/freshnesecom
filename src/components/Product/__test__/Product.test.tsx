import React from 'react';
import { screen, cleanup, waitFor } from '@testing-library/react';
import Product from '../Product';
import { ProductItem } from 'types/products';
import renderer from 'react-test-renderer';
import { RenderComponent } from 'utils/test/render';

const productMock: ProductItem = {
  brand: 'Apple',
  category: 'smartphones',
  description: 'An apple mobile which is nothing like apple',
  discountPercentage: 12.96,
  id: 1,
  images: [
    'https://dummyjson.com/image/i/products/1/1.jpg',
    'https://dummyjson.com/image/i/products/1/2.jpg',
  ],
  price: 549,
  rating: 4.69,
  stock: 94,
  thumbnail: 'https://dummyjson.com/image/i/products/1/thumbnail.jpg',
  title: 'iPhone 9',
};

function MyComponent() {
  return (
    <RenderComponent>
      <Product {...productMock} />
    </RenderComponent>
  );
}

describe('Product', () => {
  afterEach(cleanup);

  it('renders correctly', () => {
    const tree = renderer.create(<MyComponent />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
