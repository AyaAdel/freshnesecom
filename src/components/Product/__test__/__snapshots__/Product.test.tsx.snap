// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product renders correctly 1`] = `
<div
  className="product-item"
  data-testid="product1"
>
  <div
    className="product-item__image-wrapper"
  >
    <img
      className="product-item__image"
      src="https://dummyjson.com/image/i/products/1/thumbnail.jpg"
    />
    <span
      className="product-item__discount-percent"
    >
      -12.96%
    </span>
  </div>
  <h5
    className="product-item__name"
  >
    iPhone 9
  </h5>
  <p
    className="product-item__description"
  >
    An apple mobile which is nothing like apple
  </p>
  <div
    className="product-item__Action"
  >
    <div
      className="product-item__price"
    >
      <span
        className="product-item__current-price"
      >
        549 usd
      </span>
    </div>
    <button
      className="btn"
      onClick={[Function]}
    >
      Buy now
    </button>
  </div>
</div>
`;
